azure/scope/machinepoolmachine.go

```diff
@@ -261,6 +274,152 @@ func (s *MachinePoolMachineScope) UpdateStatus(ctx context.Context) error {
 	return nil
 }
 
+// CordonAndDrain will cordon and drain the Kubernetes node associated with this AzureMachinePoolMachine.
+func (s *MachinePoolMachineScope) CordonAndDrain(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "scope.MachinePoolMachineScope.CordonAndDrain")
+	defer span.End()
+
+	var (
+		nodeRef = s.AzureMachinePoolMachine.Status.NodeRef
+		node    *corev1.Node
+		err     error
+	)
+	if nodeRef == nil || nodeRef.Name == "" {
+		node, err = s.workloadNodeGetter.GetNodeByProviderID(ctx, s.ProviderID())
+	} else {
+		node, err = s.workloadNodeGetter.GetNodeByObjectReference(ctx, *nodeRef)
+	}
+
+	if err != nil && apierrors.IsNotFound(err) {
+		return nil // node was already gone, so no need to cordon and drain
+	} else if err != nil {
+		return errors.Wrap(err, "failed to find node")
+	}
+
+	// Drain node before deletion and issue a patch in order to make this operation visible to the users.
+	if s.isNodeDrainAllowed() {
+		patchHelper, err := patch.NewHelper(s.AzureMachinePoolMachine, s.client)
+		if err != nil {
+			return errors.Wrap(err, "failed to build a patchHelper when draining node")
+		}
+
+		s.V(4).Info("Draining node", "node", s.AzureMachinePoolMachine.Status.NodeRef.Name)
+		// The DrainingSucceededCondition never exists before the node is drained for the first time,
+		// so its transition time can be used to record the first time draining.
+		// This `if` condition prevents the transition time to be changed more than once.
+		if conditions.Get(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition) == nil {
+			conditions.MarkFalse(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition, clusterv1.DrainingReason, clusterv1.ConditionSeverityInfo, "Draining the node before deletion")
+		}
+
+		if err := patchHelper.Patch(ctx, s.AzureMachinePoolMachine); err != nil {
+			return errors.Wrap(err, "failed to patch AzureMachinePoolMachine")
+		}
+
+		if err := s.drainNode(ctx, node); err != nil {
+			conditions.MarkFalse(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition, clusterv1.DrainingFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+			return err
+		}
+
+		conditions.MarkTrue(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition)
+	}
+
+	return nil
+}
+
+func (s *MachinePoolMachineScope) drainNode(ctx context.Context, node *corev1.Node) error {
+	ctx, span := tele.Tracer().Start(ctx, "scope.MachinePoolMachineScope.drainNode")
+	defer span.End()
+
+	restConfig, err := remote.RESTConfig(ctx, MachinePoolMachineScopeName, s.client, client.ObjectKey{
```

Comment:
can we refactor to use workloadClusterProxy's client here?

Reply:
Sure. I just lifted this code from the CAPI implementation. I didn't put much work into this section as I think https://github.com/kubernetes-sigs/cluster-api/issues/4063 will end up superseding this behavior.

Reply:
`framework.ClusterProxy` is part of the test framework. I don't think we should have a dependency on the test frame work in the infra provider.

Reply:
Maybe I misunderstood, I was referring to [workloadClusterProxy](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/128493dd756748e9d3efb0bd274064de48491a86/azure/scope/machinepoolmachine.go#L49) in MachinePoolMachineScope. I see that it already has a client for the workload cluster. So, I was wondering, maybe we can update the NodeGetter interface to get the client as well.  I don't think it's a big deal though, fine with creating a followup if it makes sense.

Reply:
I think I misunderstood. I think I understand a little better now. I dried up the `getWorkloadClient` func to use `remote.NewClusterClient` which then depends on `remote.RESTConfig`.

I don't think we can use the remote client in the `workloadClusterProxy` because we need to construct `k8s.io/client-go/kubernetes.Interface` for: 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1435/files#diff-52e6076fd15ed8417a4824b81a78ef786b227161e9ad53fbe62268f21d7a3a3bR354